README.md
# ![sly logo](https://raw.githubusercontent.com/dunkyl/SlyMeta/main/sly%20logo.svg) Sly YouTube Data API for Python

<!-- elevator begin -->

> 🚧 **This library is an early work in progress! Breaking changes may be frequent.**

> 🐍 For Python 3.10+

No-boilerplate, async and typed YouTube Data API access. 😋

```shell
pip install slyytdapi
```

This library does not have full coverage.
All methods are read-only.
Currently, the following topics are supported:

* Videos
* Channels
* Comment threads
* Video search
* Channel members (requires approval from YouTube)

For collecting statistics about your own channel using the YouTube analytics API, see [YTAAPI](https://github.com/dunkyl/SlyYTAAPI-Python).

<!-- elevator end -->

---

Example usage:

```python
import asyncio
from SlyYTDAPI import *

async def main():
    # don't forget to keep your secrets secret!
    yt = YouTubeData(open('api_key.txt').read())

    my_video = await yt.video('dQw4w9WgXcQ')
    print(F"Check this out!\n{my_video.link()}")

    # keep it simple
    _ = await my_video.comments(limit=10) # list[Comment]

    # or opt in to generators
    print('\n---\n'.join([
        F"{c.author_name} > {c.body}"
        async for c in my_video.comments(limit=10)
    ]))
    
asyncio.run(main())
```

---

Example CLI usage for getting authorized:

```sh
# WINDOWS
py -m SlyYTDAPI grant
# MacOS or Linux
python3 -m SlyYTDAPI grant
```

Granting credentials requires a Google Cloud Console account and JSON file.
Please see https://docs.dunkyl.net/SlyAPI-Python/tutorial/oauth2.html for more information.


CHANGELOG.md
# Changelog

## [Unreleased]

---

## [0.3.1] - 2023-07-30

### Fixed

- `Comment.author_display_name` is now set correctly

## [0.3.0] - 2023-07-30

Since the last release, YouTube introduced a new username system.
It is important to note that `Comment` in most cases does **not** have the @username of the commenter, and it must instead be retrieved using `YouTubeData.channels` (for many channels at once) or `YouTubeData.channel` (for just one channel at a time).

Commenters are still uniquely identified by `Comment.author_channel_id`.

### Changed
- Reflect YouTube changes to usernames
    - `Channel.name` renamed to `Channel.display_name`
    - `Comment.author_name` renamed to `Comment.author_display_name`
    - `Channel.custom_url` renamed `Channel.at_username` and is never None
    - deprecated properties to access these by their old name

### Fixed
- `Channel.created_at` deserialization format now accepts seconds fraction
- `YouTubeData.channels` no longer bad request when using >50 channel IDs

## [0.2.0] - 2023-03-01

### Changed
- `YouTubeData_WithMembers.get_my_levels` does not accept a parts parameter and always uses ID and SNIPPET
- Updated for SlyAPI 0.4.3
- `YouTubeData_WithMembers` constructor now takes only `OAuth2` 
- `YouTubeData` and `YouTubeData_WithMembers` should no longer be awaited
- `Part`, `PrivacyStatus`, `SafeSearch`, `Order`, and `CommentOrder` are now plain enums
- Many methods now take `T|set[T]` for enum parameters

## [0.1.3] - 2022-02-26

### Changed
- Updated for SlyAPI 0.2.4

## [0.1.2] - 2022-02-23

### Changed
- Updated for SlyAPI 0.2.0

## [0.1.1] - 2022-02-13

### Fixed
- `YouTubeData` constructor
- `Video.link()` if video came from search results

## [0.1.0] - 2021-11-24

### Changed
- Updated for SlyAPI 0.1.0

## [0.0.3] - 2021-11-22

Initial release.

test/search.py
import asyncio
from SlyYTDAPI import *

async def main():
    yt = await YouTubeData_WithMembers(open('api_key.txt').read())

    search_results = await yt.search_videos('gangnam style', limit=5)

    for video in search_results:
        print(F"{video.title} by {video.channel_name}")

asyncio.run(main())

test/members.py
import asyncio
from SlyYTDAPI import *

async def main():
    yt = await YouTubeData_WithMembers(open('api_key.txt').read())

    members = await yt.get_my_members()

    for member in members:
        print(member.channel_name)

asyncio.run(main())

test/channels.py
import asyncio
from SlyYTDAPI import *

async def main():
    yt = await YouTubeData_WithMembers(open('api_key.txt').read())

    channel = await yt.channel('UCy0tKL1T7wFoYcxCe0xjN6Q')

    print(channel.name)

asyncio.run(main())

src/SlyYTDAPI/members.py
'''
YouTube Members Endpoints for the Data API v3
https://developers.google.com/youtube/v3/docs/members
'''
from datetime import datetime
from enum import Enum
from typing import Any, TypedDict, cast
from SlyAPI.oauth2 import OAuth2
from SlyAPI import *
from SlyAPI.web import JsonMap
from .ytdapi import YouTubeData, Part, yt_date

class _MembersPollResponse(TypedDict):
    kind: str
    etag: str
    nextPageToken: str
    pageInfo: dict[str, int]
    items: list[dict[str, Any]]

class _MembersLevelsResponse(TypedDict):
    kind: str
    etag: str
    items: list[dict[str, Any]]

class MembersMode(Enum):
    ALL_CURRENT = 'all_current'
    UPDATES = 'updates'

class MemberLevel:
    # part: id
    id: str
    # part: snippet
    name: str

    def __init__(self, source: JsonMap):
        match source:
            case { # from /membershipsLevels endpoint
                'id': str(id),
                'snippet': {
                    'levelDetails': {
                        'displayName': str(name)
                    } } }:
                self.id = id
                self.name = name
            case { # as part of Member resource
                'highestAccessibleLevel': str(id),
                'highestAccessibleLevelDisplayName': str(name)
            }:
                self.id = id
                self.name = name
            case _:
                raise ValueError(f'Invalid source: {source}')

class Membership:
    # part: snippet
    channel_id: str
    channel_name: str
    profile_image_url: str
    level: MemberLevel
    since: datetime
    total_months: int
    since_at_level: datetime
    total_months_at_level: int

    def __init__(self, source: dict[str, Any]):

        snippet = source['snippet']
        self.channel_id = snippet['memberDetails']['channelId']
        self.channel_name = snippet['memberDetails']['displayName']
        self.profile_image_url = snippet['memberDetails']['profileImageUrl']
        self.level= MemberLevel(snippet['membershipsDetails'])
        self.since = yt_date(snippet['membershipsDuration']['memberSince'])
        self.total_months = snippet['membershipsDuration']['memberTotalDurationMonths']
        self.since_at_level = yt_date(snippet['membershipsDurationAtLevel']['memberSince'])
        self.total_months_at_level = snippet['membershipsDurationAtLevel']['memberTotalDurationMonths']
        
class YouTubeData_WithMembers(YouTubeData):

    _next_page: str|None = None

    def __init__(self, auth: OAuth2) -> None:
        super().__init__(auth)
    
    def get_my_members(self,
        level_id: str|None=None,
        member_channel_ids: list[str]|None=None,
        parts: Part|set[Part]=Part.SNIPPET,
        limit: int|None=None) -> AsyncTrans[Membership]:
        if member_channel_ids is not None and len(member_channel_ids) > 100:
            raise ValueError('Cannot fetch more than 100 specific members.')
        mode = MembersMode.ALL_CURRENT
        params = {
            'part': parts,
            'mode': mode,
            'hasAccessToLevel': level_id,
            'filterByMemberChannelId': ','.join(member_channel_ids or []),
            'maxResults': 1000 if limit is None else min(1000, limit)
        }
        return self.paginated(
            '/members', params, limit
            ).map(Membership)
    
    async def _members_poll(self, pageToken: str|None) -> _MembersPollResponse:
        params = {
            'part': Part.SNIPPET,
            'membersMode': MembersMode.UPDATES,
            'pageToken': pageToken
        }
        return cast(_MembersPollResponse, await self.get_json('/members', params))
    
    async def _memberships_levels(self, parts: Part|set[Part]) -> _MembersLevelsResponse:
        params = { 'part': parts }
        return cast(_MembersLevelsResponse, await self.get_json('/membershipsLevels', params))

    async def poll_new_members(self) -> list[Membership]:
        '''Polls for new members since the last call to this method.'''
        if self._next_page is None:
            # first call for 'updates' mode does not return any members
            # but it does return a nextPageToken always
            self._next_page = (await self._members_poll(None))['nextPageToken']
            return []
        else:
            response = await self._members_poll(self._next_page)
            self._next_page = response['nextPageToken']
            return [Membership(r) for r in response['items']]

    async def get_my_levels(self) -> list[MemberLevel]:
        return [MemberLevel(r) for r in (await self._memberships_levels({Part.ID,Part.SNIPPET}))['items']]

src/SlyYTDAPI/ytdapi.py
import re
from enum import Enum
from datetime import datetime
from typing import TypeVar, Any
from warnings import warn
from SlyAPI import *

SCOPES_ROOT = 'https://www.googleapis.com/auth/youtube'

class Scope:
    READONLY     = F"{SCOPES_ROOT}.readonly"
    MEMBERS      = F"{SCOPES_ROOT}.channel-memberships.creator"

class Part(Enum):
    ID           = 'id'             # quota cost: 0
    DETAILS      = 'contentDetails' # quota cost: 2
    SNIPPET      = 'snippet'        # quota cost: 2
    STATUS       = 'status'         # quota cost: 2
    STATISTICS   = 'statistics'     # quota cost: 2
    REPLIES      = 'replies'        # quota cost: 2
    # ...

class PrivacyStatus(Enum):
    PRIVATE      = 'private'
    UNLISTED     = 'unlisted'
    PUBLIC       = 'public'

class SafeSearch(Enum):
    SAFE         = 'strict'
    MODERATE     = 'moderate'
    UNSAFE       = 'none'

class Order(Enum):
    DATE         = 'date'
    LIKES        = 'rating'
    RELEVANCE    = 'relevance'
    ALPHABETICAL = 'title'
    VIEWS        = 'viewCount'

class CommentOrder(Enum):
    RELEVANCE    = 'relevance'
    TIME         = 'time'

ISO8601_PERIOD = re.compile(r'P(\d+)?T(?:(\d{1,2})H)?(?:(\d{1,2})M)?(\d{1,2})S')

def yt_date(date: str) -> datetime:
    try:
        return datetime.strptime(date, '%Y-%m-%dT%H:%M:%S.%fZ')
    except ValueError:
        return datetime.strptime(date, '%Y-%m-%dT%H:%M:%SZ')

W = TypeVar('W')
T = TypeVar('T')

def get_dict_path(d: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        if key not in d:
            return None
        d = d[key]
    return d
    
class Comment:
    id: str
    # part: snippet
    author_display_name: str
    author_channel_id: str
    body: str
    created_at: datetime
    # part: replies
    replies: list['Comment']|None

    @property
    def author_name(self):
        warn("author_name is deprecated, please use author_display_name")
        return self.author_display_name

    def __init__(self, source: dict[str, Any]):
        # case of top-level comment
        if tlc := source.get('snippet', {}).get('topLevelComment'):
            replies: list[Any] = source.get('replies', {}).get('comments', [])
            self.replies = [Comment(r) for r in replies]
            source = tlc
            
        self.id = source['id']
        if snippet := source.get('snippet'):
            self.author_display_name = snippet['authorDisplayName']
            self.author_channel_id = snippet['authorChannelId']['value']
            self.body = snippet['textDisplay']
            self.created_at = yt_date(snippet['publishedAt'])
        


class Video:
    _youtube: 'YouTubeData'
    id: str

    # part: snippet
    title: str
    description: str
    published_at: datetime
    channel_id: str
    channel_name: str
    tags: list[str]
    is_livestream: bool

    # part: contentDetails
    duration: int

    # part: status
    privacy: PrivacyStatus

    # part: statistics
    view_count: int
    like_count: int

    # dislike_count: int ## rest in peace
    comment_count: int

    def __init__(self, source: dict[str, Any], yt: 'YouTubeData'):
        self._youtube = yt
        match source['id']:
            case str():
                self.id = source['id']
            case dict(): # case for video search result object
                self.id = source['id']['videoId']
            case _:
                raise ValueError("Video expects source id to be a string or dict")
        
        if snippet := source.get('snippet'):
            self.title = snippet['title']
            self.description = snippet['description']
            self.published_at = yt_date(snippet['publishedAt'])
            self.channel_id = snippet['channelId']
            self.channel_name = snippet['channelTitle']
            self.tags = snippet.get('tags', [])
            self.is_livestream = snippet.get('liveBroadcastContent') == 'live'
        if contentDetails := source.get('contentDetails'):
            m = ISO8601_PERIOD.match(contentDetails['duration'])
            if m:
                days, hours, minutes, seconds = (int(g) if g else 0 for g in m.groups())
                self.duration = days * 24 * 60 * 60 + hours * 60 * 60 + minutes * 60 + seconds
            else:
                raise ValueError(F"Unknown duration format: {contentDetails['duration']}")
        if status := source.get('status'):
            self.privacy = status['privacyStatus']


    def link(self, short: bool = False) -> str:
        if not short:
            return F"https://www.youtube.com/watch?v={self.id}"
        else:
            return F"https://youtu.be/{self.id}"

    def comments(self, limit: int | None = 100) -> AsyncTrans[Comment]:
        return self._youtube.comments(self.id, limit=limit)

    async def channel(self) -> 'Channel':
        return await self._youtube.channel(self.channel_id)

class Playlist:
    _youtube: 'YouTubeData'
    id: str

    def __init__(self, id: str, yt: 'YouTubeData'):
        self._youtube = yt
        self.id = id

    def link(self) -> str:
        return F"https://www.youtube.com/playlist?list={self.id}"

class Channel:
    _youtube: 'YouTubeData'
    id: str

    # part: snippet
    display_name: str
    description: str
    created_at: datetime
    at_username: str
    profile_image_url: str|None = None

    # part: contentDetails
    uploads_playlist: Playlist

    # part: statistics
    view_count: int
    subscriber_count: int
    video_count: int

    @property
    def custom_url(self):
        warn("custom_url is deprecated, please use at_username")
        return self.at_username

    
    @property
    def name(self):
        warn("name is deprecated, please use display_name")
        return self.display_name

    def __init__(self, source: dict[str, Any], yt: 'YouTubeData'):
        self._youtube = yt

        self.id = source['id']
        if snippet := source.get('snippet'):
            self.display_name = snippet['title']
            self.description = snippet['description']
            self.created_at = yt_date(snippet['publishedAt'])
            self.profile_image_url = snippet.get('thumbnails', {}).get('default', {}).get('url')
            self.at_username = snippet.get('customUrl')

        if details := source.get('contentDetails'):
            self.uploads_playlist = Playlist(details['relatedPlaylists']['uploads'], yt)

        if stats := source.get('statistics'):
            self.view_count = int(stats['viewCount'])
            self.subscriber_count = int(stats['subscriberCount'])
            self.video_count = int(stats['videoCount'])

    def link(self) -> str:
        if self.at_username:
            return F"https://www.youtube.com/c/{self.at_username}"
        else:
            return F"https://www.youtube.com/channels/{self.id}"

    async def update(self):
        new = await self._youtube.channel(self.id)
        self.__dict__.update(new.__dict__)

    def videos(self, limit: int|None=None, mine: bool|None=None) -> AsyncTrans[Video]:
        return self._youtube.search_videos(channel_id=self.id, limit=limit, mine=mine)

class YouTubeData(WebAPI):
    base_url = 'https://www.googleapis.com/youtube/v3'

    def __init__(self, app_or_api_key: str|OAuth2|UrlApiKey) -> None:
        match app_or_api_key:
            case str():
                auth = UrlApiKey('key', app_or_api_key)
            case _:
                auth = app_or_api_key
        super().__init__(auth)

    async def my_channel(self, parts: Part=Part.SNIPPET) -> Channel:
        return (await self._channels_list(mine=True, parts=parts, limit=1))[0]

    async def channels(self, channel_ids: list[str], parts: Part) -> list[Channel]:
        return await self._channels_list(channel_ids=channel_ids, parts=parts)

    async def channel(self, channel_id: str, parts: Part=Part.SNIPPET) -> Channel:
        return (await self._channels_list(channel_ids=[channel_id], parts=parts))[0]

    def _channels_list(self,
        channel_ids: list[str]|None=None,
        mine: bool=False,
        parts: Part|set[Part]=Part.SNIPPET,
        limit: int|None=None) -> AsyncTrans[Channel]:
        if mine==bool(channel_ids):
            raise ValueError("Must specify exactly one of channel id or mine in channel list query")
        maxResults = min(50, limit) if limit else None # per-page limit
        params = { 'part': parts, 'maxResults': maxResults }
        if channel_ids:
            channel_ids = list(set(channel_ids or [])) # deduplicate IDs
            channels_chunks50 = [
                channel_ids[i: i + 50] for i in range(0, len(channel_ids), 50)
            ]
            async def page_chunks():
                for ids in channels_chunks50:
                    p = params | { 'id': ','.join(ids) }
                    async for c in self.paginated('/channels', p, limit):
                        yield c
            return AsyncLazy(page_chunks()).map(lambda r: Channel(r, self))
        else: # mine
            return self.paginated(
                '/channels', params | { 'mine': True }, limit
                ).map(lambda r: Channel(r, self))

    def videos(self,
        video_ids: list[str],
        parts: Part|set[Part]={Part.ID,Part.SNIPPET}) -> AsyncTrans[Video]:
        params = {
            'part': parts,
            'id': ','.join(video_ids),
        }
        return self.paginated(
            '/videos', params, None
            ).map(lambda r: Video(r, self))

    async def video(self, id: str, parts: Part|set[Part]={Part.ID,Part.SNIPPET}) -> Video:
        return (await self.videos([id], parts))[0]

    def get_playlist_videos(self,
        playlist_id: str, 
        parts: Part|set[Part]=Part.SNIPPET,
        limit: int|None=None) -> AsyncTrans[Video]:
        params = {
            'part': parts,
            'playlistId': playlist_id,
            'maxResults': min(50, limit) if limit else None,
        }
        return self.paginated(
            '/playlistItems', params, limit
            ).map(lambda r: Video(r, self))

    def search_videos(self,
        query: str|None=None,
        channel_id: str|None=None,
        after: datetime|None=None,
        before: datetime|None=None,
        mine: bool|None=None,
        order: Order=Order.RELEVANCE,
        safeSearch: SafeSearch=SafeSearch.MODERATE,
        parts: Part|set[Part]=Part.SNIPPET,
        limit: int|None=50) -> AsyncTrans[Video]:
        params = {
            'part': parts,
            'safeSearch': safeSearch,
            'order': order,
            'type': 'video',
            'q': query,
            'channelId': channel_id,
            'forMine': mine,
            'publishedAfter': after.isoformat("T") + "Z" if after else None,
            'publishedBefore': before.isoformat("T") + "Z" if before else None,
            'maxResults': min(50, limit) if limit else None,
        }

        return self.paginated(
            '/search', params, limit
            ).map(lambda r: Video(r, self))

    def comments(self,
        video_id: str,
        query: str|None=None,
        parts: Part|set[Part]={Part.SNIPPET,Part.REPLIES},
        order: CommentOrder=CommentOrder.TIME,
        limit: int|None=None) -> AsyncTrans[Comment]:
        params = {
            'part': parts,
            'commentOrder': order,
            'searchTerms': query,
            'videoId': video_id,
            'maxResults': min(100, limit) if limit else None,
        }
        return self.paginated(
            '/commentThreads', params, limit
            ) .map(Comment)
    